prov/rxm, verbs, util: serialization fixes + refactoring

a-ilango opened the pull request.

swelch commented on prov/verbs/src/verbs_ep.c:
```diff
@@ -240,10 +240,16 @@ static int fi_ibv_ep_close(fid_t fid)
 	case FI_EP_MSG:
+		if (ep->eq)
+			fastlock_acquire(&ep->eq->lock);
+
```
This will cause a deadlock with XRC when a message endpoint is closed (if it was not ever full connected, like a simultaneous reject).

swelch replied:
To avoid the deadlock for XRC codepath it will make more sense to take the lock inside of fi_ibv_ep_xrc_close() right before the call to fi_ibv_ep_destroy_xrc_qp(), and release it before fi_ibv_ep_xrc_close() returns.

swelch commented on prov/verbs/src/verbs_ep.c:
```diff
@@ -240,10 +240,16 @@ static int fi_ibv_ep_close(fid_t fid)
 			fi_ibv_ep_xrc_close(ep);
 		else
 			rdma_destroy_ep(ep->id);
+
```
This will cause a deadlock with XRC when a message endpoint is closed (if it was not ever fully connected, like a simultaneous RXM reject).

To avoid the deadlock for XRC codepath it will make more sense to take the lock inside of fi_ibv_ep_xrc_close() right before the call to fi_ibv_ep_destroy_xrc_qp(), and release it before fi_ibv_ep_xrc_close() returns.

swelch replied:
@a-ilango - Once the changes I have in PR #5016 are accepted, this lock can be taken where you originally have it (around the if/else clause). I included the locking you have here in PR #5016 and made the necessary changes since the XRC EP close needed to be protected in my changes as well.

a-ilango replied:
Sounds good. I'll take a look at #5016 and let's merge that PR before this one.

shefty commented on prov/rxm/src/rxm_init.c:
```diff
@@ -177,6 +177,14 @@ int rxm_info_to_rxm(uint32_t version, const struct fi_info *core_info,
+
+	return (fi_param_get_str(&rxm_prov, "default_data_progress", &param) ?
+		0 : !strcmp(param, "auto"));
+}
```
This isn't needed.  The app already has a way to specify what it requires through fi_getinfo.  If the app doesn't, then the provider should simply report what is optimal for it.  We shouldn't need another control for what's already controllable from the API directly.

a-ilango replied:
I agree but this is mostly for testing purpose. Since RxM returns only manual progress mode if the app doesn't explicitly specify one in hints, it's hard to test auto-progress for apps like MPI which leave the progress mode undefined in hints. Some MPI may have run-time options that set the progress mode but having an env variable in rxm makes it a one-stop solution.

shefty replied:
We don't want to add environment variables unless they are absolutely needed.  It's just more code that needs to be maintained and tests.  (Yes, we should validate the env vars do what they are supposed to).  Please drop this patch.

shefty commented on prov/rxm/src/rxm_init.c:
```diff
@@ -106,7 +106,10 @@ int rxm_info_to_core(uint32_t version, const struct fi_info *hints,
+			if (rxm_needs_atomic_progress(hints))
+				core_info->domain_attr->threading = FI_THREAD_SAFE;
+			else
+				core_info->domain_attr->threading = hints->domain_attr->threading;
```
This is reverting a change made about 2 commits lower in the stack.  Rather than applying a revert commit, I would simply make the changes that are needed.  If this change is needed, then remove it from the prior commit, along with the description from the patch message.

a-ilango replied:
Hopefully my response to previous comment makes it clear.

shefty commented on prov/rxm/src/rxm_conn.c:
```diff
@@ -1462,15 +1462,9 @@ static int rxm_conn_atomic_progress_eq_cq(struct rxm_ep *rxm_ep,
-
-		if (rxm_ep->msg_info->domain_attr->threading != FI_THREAD_SAFE)
-			rxm_ep->cmap->release(&rxm_ep->cmap->lock);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
```
This is changing to the ep lock, so the commit message is confusing.  A prior commit already killed cmap lock, didn't it?

a-ilango replied:
Ah, Github rearranged the commit order based on commit date instead of the real order. It always does this. :/

I can modify the dates if it makes it easier to review. Here is the correct order:

```
67225c8 prov/rxm: add an env variable that defines the default data progress mode
2999cc4 prov/rxm: fix missing error handling for FI_GETWAIT
6c07329 prov/verbs: serialize rdma_cm_get_event and rdma_destroy_ep
d887bd5 prov/rxm: add a unspec CM thread notification event
644254a prov/util: define new ep_list_lock for accessing av->ep_list
4d9130d Revert "prov/rxm: fix serialization for atomics auto-progress"
e9e15ea prov/rxm: replace all cmap->locks with ep->lock and add any missing serialization
3218321 prov/rxm: reject connection request in shutdown state
eb3cd74 prov/rxm: improve logging + debug asserts
```

shefty replied:
I'm going to see if I can file a bug against github for this.  It makes it difficult to review PRs by looking at individual commits.

shefty replied:
github knows this.  They just don't care.  There's a work-around for this, but it's a pain.  The problem shows up when doing git rebase.

a-ilango replied:
What's the workaround? Even if we can detect that the chronological and topological order don't match, we can manually correct it.

The issue is a minor annoyance but it can catch people off-guard.

shefty replied:
I consider it more than a minor annoyance.  :)  See this link for a work-around.  It's a manual thing that everyone would need to do when running git rebase and updating a PR.

https://stackoverflow.com/questions/44704519/how-to-fix-commit-order-in-github-pull-requests-broken-by-git-rebase

vkrishna commented on prov/rxm/src/rxm_ep.c:
```diff
@@ -2382,8 +2382,10 @@ static int rxm_listener_open(struct rxm_ep *rxm_ep)
-		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL, "Unable to get MSG EQ fd\n");
+	if (ret) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL, "unable to get MSG EQ fd\n");
+		goto err;
```
I believe this has always been failing on macOS. But now its visible.

shefty replied:
Is this something that can be fixed in the tcp provider or util code?  Is this from epoll emulation on OS X?

shefty replied:
Looking at the rxm code, I don't see a strong reason why the msg_eq_fd is acquired here.  The msg_eq_fd is needed when binding a CQ that is associated with a wait set to the EP, or when doing something with atomic progress (rxm_conn_atomic_progress_eq_cq()).  The latter case tries to get the msg_eq_fd independently.

If we're not dealing with either of those cases, then the msg_eq_fd isn't needed.  The fi_control() call can be moved out of here and into rxm_ep_wait_fd_add.  That way, we only try to access it if it's actually needed.

The msg_cq_fd has a similar access pattern.  It is needed with atomic progress (rxm_conn_atomic_progress_eq_cq()), but unlike the msg_eq_fd, that function does not call fi_control() to get it and assumes it's already available.  It is also needed by rxm_ep_wait_fd_add().

I suggest modifying rxm to have a single call that retrieves both the msg_eq_fd and msg_cq_fd, which is called from rxm_conn_atomic_progress_eq_cq() and rxm_ep_wait_fd_add().  I don't think we to store either fd with the rxm_ep.  The fd's could be obtained and used directly in the two functions that use them.

vkrishna replied:
FI_GETWAIT implies wait object of type FI_WAIT_FD to be return by util code. But providing the emulated epoll fd won't help because calling system level poll calls on it won't detect the changes on the fds that are represented by emulated epoll fd. 

So may be using kqueue in this case would help for macOS.

a-ilango replied:
Thanks @vkrishna for checking this. I'll update the code based on @shefty 's suggestion.

Are tests that use CQ wait objects excluded for tcp provider on Mac OS?

vkrishna replied:
Yes

a-ilango commented:
Update:
- removed commit that added env variable for progress
- added a commit that refactors msg CQ open and getting msg_eq and msg_cq fds.
- added EP bind function to util code that accepts ep_fid
- updated the verbs EQ serialization commit to cover more cases
- fixed commit order

shefty commented on prov/util/src/util_av.c:
```diff
@@ -509,6 +509,7 @@ int ofi_av_init_lightweight(struct util_domain *domain, const struct fi_av_attr
 	 */
 	av->context = context;
 	av->domain = domain;
+	fastlock_init(&av->ep_list_lock);
```
Is this destroyed somewhere?

a-ilango replied:
Thanks. Fixed it.
